ow_power_system/src/PowerSystemNode.cpp

```diff
@@ -263,30 +268,155 @@ void PowerSystemNode::injectFault (const string& fault_name,
     {
       ros::param::getCached("/faults/" + FAULT_NAME_HPD, hpd_wattage);
       wattage += hpd_wattage;
-    } else
+    }
+  }
+}
+
+void PowerSystemNode::injectCustomFault(bool& fault_activated,
+                                        const PrognoserVector& sequence,
+                                        size_t& index,
+                                        double& wattage,
+                                        double& voltage,
+                                        double& temperature)
+{
+  static string saved_fault_directory = "N/A";
+  string current_fault_directory;
+  string designated_file;
+  static string saved_file;
+  static bool custom_fault_ready = false;
+  static bool custom_warning_displayed = false;
+  static bool end_fault_warning_displayed = false;
+  bool fault_enabled = false;
+
+  // Do nothing unless the specified fault has been injected.
+  if (! ros::param::getCached("/faults/activate_custom_fault", fault_enabled)) {
+    return;
+  }
+
+  if (!fault_activated && fault_enabled)
+  {
+    // Only continue if no other faults have been injected.
+    ROS_INFO_STREAM("Attempting custom fault activation...");
+    if (HPD_FAULT_ACTIVATED)
+    {
+      if (!custom_warning_displayed)
+      {
+        ROS_WARN_STREAM("Cannot activate custom fault: high power draw already used this simulation.");
+        custom_warning_displayed = true;
+      }
+    }
+    else
+    {
+      // Get user-entered file directory
+      if (! ros::param::getCached("/faults/custom_fault_profile", designated_file))
+      {
+        ROS_ERROR("Failed to fetch custom fault name.");
+        return;
+      }
+
+      // Append the current fault directory to the stored file path.
+      current_fault_directory = ros::package::getPath("ow_power_system") + "/profiles/" + designated_file;
+      
+      if (current_fault_directory != saved_fault_directory || end_fault_warning_displayed)
+      {
+        // It's a different fault profile from whatever the previous stored directory was, OR the
+        // previous fault ran its course, so reset index.      
+        // Load a new fault profile from the provided directory.
+        try
+        {
+          ifstream file(current_fault_directory);
+          if (current_fault_directory.substr(current_fault_directory.size() - 4) != ".csv")
```

Comment:
Minor: There may be an API for getting the extension from a pathname, but this is okay and clear.

Reply:
It's definitely a bit of a hack, but since we are only looking for a single type of file extension, I figured it wasn't a top priority.

Reply:
That's fine, you can leave it like this.